graphql-dgs-subscriptions-sse/src/main/kotlin/com/netflix/graphql/dgs/subscriptions/sse/DgsSSESubscriptionHandler.kt

```diff
@@ -44,119 +49,74 @@ import java.util.*
-            subscriber.subscription.cancel()
-        }
-
-        emitter.onTimeout {
-            logger.warn("Subscription {} timed out", sessionId)
-            subscriber.subscription.cancel()
+            logger.error(errorMessage)
+            throw ServerWebInputException(errorMessage)
         }
 
         val publisher = try {
             executionResult.getData<Publisher<ExecutionResult>>()
-        } catch (ex: ClassCastException) {
-            return if (query.contains("subscription")) {
-                logger.error("Invalid return type for subscription datafetcher. A subscription datafetcher must return a Publisher<ExecutionResult>. The query was $query", ex)
-                emitter.send("Invalid return type for subscription datafetcher. Was a non-subscription query send to the subscription endpoint?")
-                emitter.complete()
-                ResponseEntity.status(500).body(emitter)
-            } else {
-                logger.warn("Invalid return type for subscription datafetcher. The query sent doesn't appear to be a subscription query: $query", ex)
-                emitter.send("Invalid return type for subscription datafetcher. Was a non-subscription query send to the subscription endpoint?")
-                emitter.complete()
-                ResponseEntity.badRequest().body(emitter)
-            }
+        } catch (exc: ClassCastException) {
+            logger.error(
+                "Invalid return type for subscription datafetcher. A subscription datafetcher must return a Publisher<ExecutionResult>. The query was {}",
+                query, exc
+            )
+            throw ServerErrorException("Invalid return type for subscription datafetcher. Was a non-subscription query send to the subscription endpoint?", exc)
         }
 
-        publisher.subscribe(subscriber)
+        val subscriptionId = UUID.randomUUID().toString()
+        return Flux.from(publisher)
+            .contextWrite { ctx -> ctx.put("subscriptionId", subscriptionId) }
+            .map {
+                val payload = SSEDataPayload(data = it.getData(), errors = it.errors, subId = subscriptionId)
+                ServerSentEvent.builder(mapper.writeValueAsString(payload))
+                    .id(UUID.randomUUID().toString())
+                    .build()
+            }.onErrorResume { exc ->
+                logger.warn("An exception occurred on subscription {}", subscriptionId, exc)
+                val errorMessage = exc.message ?: "An exception occurred"
+                val payload = SSEDataPayload(data = null, errors = listOf(SseError(errorMessage)), subId = subscriptionId)
+                Flux.just(
+                    ServerSentEvent.builder(mapper.writeValueAsString(payload))
+                        .id(UUID.randomUUID().toString())
+                        .build()
+                )
+            }
+    }
 
-        return ResponseEntity.ok(emitter)
+    private fun isSubscriptionQuery(query: String): Boolean {
```

Comment:
I changed the endpoint to explicitly check if the query is a subscription query before attempting to execute it, which seemed like a good idea to me.